File: tasks-core-jvm/src/main/java/org/funfix/tasks/jvm/DelayedFun.java
package org.funfix.tasks.jvm;

import org.jspecify.annotations.NullMarked;

/**
 * Represents a delayed computation (a thunk).
 * <p>
 * These functions are allowed to trigger side effects, with
 * blocking I/O and to throw exceptions.
 */
@NullMarked
@FunctionalInterface
public interface DelayedFun<T> {
    T invoke() throws Exception;
}

File: tasks-core/src/commonMain/kotlin/org/funfix/tasks/DelayedFun.kt
@file:Suppress("EXPECT_ACTUAL_CLASSIFIERS_ARE_IN_BETA_WARNING")

package org.funfix.tasks

import org.funfix.tasks.support.Blocking

/**
 * Represents a delayed computation (a thunk).
 *
 * These functions are allowed to trigger side effects, with
 * blocking I/O and to throw exceptions.
 */
public expect fun interface DelayedFun<out T> {
    @Blocking
    public operator fun invoke(): T
}
